// skinview3d/libs/viewer.js
import { inferModelType, isTextureSource, loadCapeToCanvas, loadImage, loadSkinToCanvas } from "skinview-utils";
import { Color, PointLight, EquirectangularReflectionMapping, Group, NearestFilter, PerspectiveCamera, Scene, Texture, Vector2, WebGLRenderer, AmbientLight } from "three";
import { RootAnimation } from "./animation.js";
import { PlayerObject } from "./model.js";
export class SkinViewer {
    constructor(options = {}) {
        this.animations = new RootAnimation();
        this.globalLight = new AmbientLight(0xffffff, 0.4);
        this.cameraLight = new PointLight(0xffffff, 0.6);
        this.backgroundTexture = null;
        this._disposed = false;
        this._renderPaused = false;
        this.canvas = options.canvas === undefined ? document.createElement("canvas") : options.canvas;
        // texture
        this.skinCanvas = document.createElement("canvas");
        this.skinTexture = new Texture(this.skinCanvas);
        this.skinTexture.magFilter = NearestFilter;
        this.skinTexture.minFilter = NearestFilter;
        this.capeCanvas = document.createElement("canvas");
        this.capeTexture = new Texture(this.capeCanvas);
        this.capeTexture.magFilter = NearestFilter;
        this.capeTexture.minFilter = NearestFilter;
        this.scene = new Scene();
        this.camera = new PerspectiveCamera();
        this.camera.add(this.cameraLight);
        this.scene.add(this.camera);
        this.scene.add(this.globalLight);
        this.renderer = new WebGLRenderer({
            canvas: this.canvas,
            alpha: options.alpha !== false,
            preserveDrawingBuffer: options.preserveDrawingBuffer === true // default: false
        });
        this.renderer.setPixelRatio(window.devicePixelRatio);
        this.playerObject = new PlayerObject(this.skinTexture, this.capeTexture);
        this.playerObject.name = "player";
        this.playerObject.skin.visible = false;
        this.playerObject.cape.visible = false;
        this.playerWrapper = new Group();
        this.playerWrapper.add(this.playerObject);
        this.scene.add(this.playerWrapper);
        if (options.skin !== undefined) {
            this.loadSkin(options.skin, options.model);
        }
        if (options.cape !== undefined) {
            this.loadCape(options.cape);
        }
        if (options.width !== undefined) {
            this.width = options.width;
        }
        if (options.height !== undefined) {
            this.height = options.height;
        }
        if (options.background !== undefined) {
            this.background = options.background;
        }
        if (options.panorama !== undefined) {
            this.loadPanorama(options.panorama);
        }
        this.camera.position.z = 1;
        this._zoom = options.zoom === undefined ? 0.9 : options.zoom;
        this.fov = options.fov === undefined ? 50 : options.fov;
        if (options.renderPaused === true) {
            this._renderPaused = true;
            this.animationID = null;
        }
        else {
            this.animationID = window.requestAnimationFrame(() => this.draw());
        }
        this.onContextLost = (event) => {
            event.preventDefault();
            if (this.animationID !== null) {
                window.cancelAnimationFrame(this.animationID);
                this.animationID = null;
            }
        };
        this.onContextRestored = () => {
            if (!this._renderPaused && !this._disposed && this.animationID === null) {
                this.animationID = window.requestAnimationFrame(() => this.draw());
            }
        };
        this.canvas.addEventListener("webglcontextlost", this.onContextLost, false);
        this.canvas.addEventListener("webglcontextrestored", this.onContextRestored, false);
    }
    loadSkin(source, model = "auto-detect", options = {}) {
        if (source === null) {
            this.resetSkin();
        }
        else if (isTextureSource(source)) {
            loadSkinToCanvas(this.skinCanvas, source);
            const actualModel = model === "auto-detect" ? inferModelType(this.skinCanvas) : model;
            this.skinTexture.needsUpdate = true;
            this.playerObject.skin.modelType = actualModel;
            if (options.makeVisible !== false) {
                this.playerObject.skin.visible = true;
            }
        }
        else {
            return loadImage(source).then(image => this.loadSkin(image, model, options));
        }
    }
    resetSkin() {
        this.playerObject.skin.visible = false;
    }
    loadCape(source, options = {}) {
        if (source === null) {
            this.resetCape();
        }
        else if (isTextureSource(source)) {
            loadCapeToCanvas(this.capeCanvas, source);
            this.capeTexture.needsUpdate = true;
            if (options.makeVisible !== false) {
                this.playerObject.backEquipment = options.backEquipment === undefined ? "cape" : options.backEquipment;
            }
        }
        else {
            return loadImage(source).then(image => this.loadCape(image, options));
        }
    }
    resetCape() {
        this.playerObject.backEquipment = null;
    }
    loadPanorama(source) {
        return this.loadBackground(source, EquirectangularReflectionMapping);
    }
    loadBackground(source, mapping) {
        if (isTextureSource(source)) {
            if (this.backgroundTexture !== null) {
                this.backgroundTexture.dispose();
            }
            this.backgroundTexture = new Texture();
            this.backgroundTexture.image = source;
            if (mapping !== undefined) {
                this.backgroundTexture.mapping = mapping;
            }
            this.backgroundTexture.needsUpdate = true;
            this.scene.background = this.backgroundTexture;
        }
        else {
            return loadImage(source).then(image => this.loadBackground(image, mapping));
        }
    }
    draw() {
        this.animations.runAnimationLoop(this.playerObject);
        this.render();
        this.animationID = window.requestAnimationFrame(() => this.draw());
    }
    /**
    * Renders the scene to the canvas.
    * This method does not change the animation progress.
    */
    render() {
        this.renderer.render(this.scene, this.camera);
    }
    setSize(width, height) {
        this.camera.aspect = width / height;
        this.camera.updateProjectionMatrix();
        this.renderer.setSize(width, height);
    }
    dispose() {
        this._disposed = true;
        this.canvas.removeEventListener("webglcontextlost", this.onContextLost, false);
        this.canvas.removeEventListener("webglcontextrestored", this.onContextRestored, false);
        if (this.animationID !== null) {
            window.cancelAnimationFrame(this.animationID);
            this.animationID = null;
        }
        this.renderer.dispose();
        this.skinTexture.dispose();
        this.capeTexture.dispose();
        if (this.backgroundTexture !== null) {
            this.backgroundTexture.dispose();
            this.backgroundTexture = null;
        }
    }
    get disposed() {
        return this._disposed;
    }
    /**
     * Whether rendering and animations are paused.
     * Setting this property to true will stop both rendering and animation loops.
     * Setting it back to false will resume them.
     */
    get renderPaused() {
        return this._renderPaused;
    }
    set renderPaused(value) {
        this._renderPaused = value;
        if (this._renderPaused && this.animationID !== null) {
            window.cancelAnimationFrame(this.animationID);
            this.animationID = null;
        }
        else if (!this._renderPaused && !this._disposed && !this.renderer.getContext().isContextLost() && this.animationID == null) {
            this.animationID = window.requestAnimationFrame(() => this.draw());
        }
    }
    get width() {
        return this.renderer.getSize(new Vector2()).width;
    }
    set width(newWidth) {
        this.setSize(newWidth, this.height);
    }
    get height() {
        return this.renderer.getSize(new Vector2()).height;
    }
    set height(newHeight) {
        this.setSize(this.width, newHeight);
    }
    get background() {
        return this.scene.background;
    }
    set background(value) {
        if (value === null || value instanceof Color || value instanceof Texture) {
            this.scene.background = value;
        }
        else {
            this.scene.background = new Color(value);
        }
        if (this.backgroundTexture !== null && value !== this.backgroundTexture) {
            this.backgroundTexture.dispose();
            this.backgroundTexture = null;
        }
    }
    adjustCameraDistance() {
        let distance = 4.5 + 16.5 / Math.tan(this.fov / 180 * Math.PI / 2) / this.zoom;
        // limit distance between 10 ~ 256 (default min / max distance of OrbitControls)
        if (distance < 10) {
            distance = 10;
        }
        else if (distance > 256) {
            distance = 256;
        }
        this.camera.position.multiplyScalar(distance / this.camera.position.length());
        this.camera.updateProjectionMatrix();
    }
    get fov() {
        return this.camera.fov;
    }
    set fov(value) {
        this.camera.fov = value;
        this.adjustCameraDistance();
    }
    get zoom() {
        return this._zoom;
    }
    set zoom(value) {
        this._zoom = value;
        this.adjustCameraDistance();
    }
}
